packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -0,0 +1,224 @@
+import type {
+  Metadata,
+  ResolvedMetadata,
+  ResolvingMetadata,
+} from './types/metadata-interface'
+import type { Viewport } from './types/extra-types'
+import type { ResolvedTwitterMetadata } from './types/twitter-types'
+import type { AbsoluteTemplateString } from './types/metadata-types'
+import { createDefaultMetadata } from './default-metadata'
+import { resolveOpenGraph } from './resolve-opengraph'
+import { mergeTitle } from './resolve-title'
+
+const viewPortKeys = {
+  width: 'width',
+  height: 'height',
+  initialScale: 'initial-scale',
+  minimumScale: 'minimum-scale',
+  maximumScale: 'maximum-scale',
+  viewportFit: 'viewport-fit',
+} as const
+
+type Item =
+  | {
+      type: 'layout' | 'page'
+      // A number that represents which layer or routes that the item is in. Starting from 0.
+      // Layout and page in the same level will share the same `layer`.
+      layer: number
+      mod: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path: string
+    }
+  | {
+      type: 'icon'
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod?: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path?: string
+    }
+
+// Merge the source metadata into the resolved target metadata.
+function merge(
+  target: ResolvedMetadata,
+  source: Metadata,
+  templateStrings: {
+    title: string | null
+    openGraph: string | null
+    twitter: string | null
+  }
+) {
+  for (const key_ in source) {
+    const key = key_ as keyof Metadata
+
+    switch (key) {
+      case 'other': {
+        Object.assign(target.other, source.other)
+        break
+      }
+      case 'title': {
+        if (source.title) {
+          target.title = source.title as AbsoluteTemplateString
+          mergeTitle(target, templateStrings.title)
```

Comment:
target.title could be `null`, so that's why we need to pass either the `target` itself, or `source` itself into the merging then assign it back to `target`. otherwise we need to give target.title an initial value, which requires to create extra object here that didn't match our expectation for memory saving?

Reply:
Oh right, but we can solve that outside of the merge call.

```
If (source.title === null) {
  target.title = source.title;
} else {
  If (target.title === null) {
    target.title =  { /* new empty title object */ }
  }
  mergeTitle(target.title, source.title, templates.title)
}
```

then you only allocate if you need an object

Reply:
yeah but we need to it for opengraph/twitter as well, feel bit duplicated? that was the purpose to encapsulate into the `mergeTitle`